File: src/main/java/com/zepben/testutils/hamcrest/InputStreamMatcher.java
/*
 * Copyright 2020 Zeppelin Bend Pty Ltd
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package com.zepben.testutils.hamcrest;

import com.zepben.annotations.EverythingIsNonnullByDefault;
import org.hamcrest.Description;
import org.hamcrest.TypeSafeMatcher;

import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.equalTo;

@SuppressWarnings("WeakerAccess")
@EverythingIsNonnullByDefault
public class InputStreamMatcher extends TypeSafeMatcher<InputStream> {

    private final String expected;

    public static InputStreamMatcher matchesContent(String expected) {
        return new InputStreamMatcher(expected);
    }

    @Override
    public void describeTo(Description description) {
        description.appendText("InputStream containing ").appendValue(expected);
    }

    @Override
    protected void describeMismatchSafely(InputStream item, Description mismatchDescription) {
        mismatchDescription.appendText(" was ").appendValue(toString(item));
    }

    @Override
    protected boolean matchesSafely(InputStream item) {
        return expected.equals(toString(item));
    }

    private String toString(InputStream inputStream) {
        try {
            int size = inputStream.available();
            byte[] bytes = new byte[size];
            assertThat(inputStream.read(bytes, 0, size), equalTo(size));
            return new String(bytes, StandardCharsets.UTF_8);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    private InputStreamMatcher(String expected) {
        this.expected = expected;
    }

}

File: src/main/kotlin/com/zepben/testutils/hamcrest/InputStreamMatcher.kt
/*
 * Copyright 2020 Zeppelin Bend Pty Ltd
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package com.zepben.testutils.hamcrest

import org.hamcrest.Description
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.equalTo
import org.hamcrest.TypeSafeMatcher
import java.io.IOException
import java.io.InputStream
import java.nio.charset.StandardCharsets

class InputStreamMatcher private constructor(private val expected: String) : TypeSafeMatcher<InputStream>() {

    override fun describeTo(description: Description) {
        description.appendText("InputStream containing ").appendValue(expected)
    }

    public override fun describeMismatchSafely(item: InputStream, mismatchDescription: Description) {
        mismatchDescription.appendText(" was ").appendValue(item.asString())
    }

    public override fun matchesSafely(item: InputStream): Boolean =
        expected == item.asString()

    private fun InputStream.asString(): String =
        try {
            val size = available()
            val bytes = ByteArray(size)

            assertThat(read(bytes, 0, size), equalTo(size))

            String(bytes, StandardCharsets.UTF_8)
        } catch (e: IOException) {
            throw RuntimeException(e)
        }

    companion object {

        @JvmStatic
        fun matchesContent(expected: String): InputStreamMatcher {
            return InputStreamMatcher(expected)
        }

    }

}
